src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs

```diff
@@ -0,0 +1,411 @@
+                    {
+                        var i = 0L;
+                        while (i++ < max && await reader.ReadAsync(cts.Token))
+                        {
+                            var id = reader.GetString(0);
+                            var sequenceNr = reader.GetInt64(1);
+                            var timestamp = reader.GetInt64(2);
+                            var isDeleted = reader.GetBoolean(3);
+                            var manifest = reader.GetString(4);
+                            var payload = reader[5];
+                            var serializerId = reader.GetInt32(6);
+                            
+                            // Deserialize the persistent payload using the data we read from the reader
+                            var deserialized = _serialization.Deserialize((byte[])payload, serializerId, manifest);
+                            
+                            // Call the recovery callback with the deserialized data from the database
+                            recoveryCallback(new Persistent(
+                                deserialized, 
+                                sequenceNr,
+                                id,
+                                manifest, 
+                                isDeleted, 
+                                ActorRefs.NoSender, 
+                                null, 
+                                timestamp));
+                        }
+                        command.Cancel();
+                    }
+                }
+            }
+        }
+
+        public sealed override async Task<long> ReadHighestSequenceNrAsync(string persistenceId, long fromSequenceNr)
+        {
+            // Create a new DbConnection instance
+            using (var connection = new SqliteConnection(_connectionString))
+            using (var cts = CancellationTokenSource.CreateLinkedTokenSource(_pendingRequestsCancellation.Token))
+            {
+                await connection.OpenAsync(cts.Token);
+                // Create new DbCommand instance
+                using (var command = GetCommand(connection, HighestSequenceNrSql, _timeout))
+                {
+                    // Populate the SQL parameters
+                    AddParameter(command, "@PersistenceId", DbType.String, persistenceId);
+        
+                    // Execute the SQL query statement
+                    var result = await command.ExecuteScalarAsync(cts.Token);
+        
+                    // Return the result if one is returned by the query result, else return zero
+                    return result is long ? Convert.ToInt64(result) : 0L;
+                }
+            }
+        }
+
+        protected sealed override async Task<IImmutableList<Exception>> WriteMessagesAsync(IEnumerable<AtomicWrite> messages)
+        {
+            // For each of the atomic write request, create an async Task 
+            var writeTasks = messages.Select(async message =>
```

Comment:
Questions:

 - What is the risk that this task spawning pattern writes messages out of order? (In theory this may not be a big deal, but could lead to interesting edge cases)
 
 - What is the risk of this pattern hammering the datasource with open connections and causing issues on scale?

Reply:
@Arkatufus when we were discussing this documentation, we had raised the issue of having a "caveats and pitfalls" section - I think these two questions definitely belong in there: thinking about operation order (how does Akka.Persistence manage this for you vs. what do you need to worry about yourself) and connection management.

The built-in Akka.Persistence fundamentals should take care of the former (`AtomicWrite` is ordered and all operations in the `AtomicWrite` should be written in the same batch) but that semantic needs to be explained so an implementer can do this correctly.

Reply:
That is an interesting question, in theory, there is no guarantee that each `AtomicWrite` Task would be written in order, especially if one of them in the middle failed to be written to the datasource, the ordering would be screwed. @Aaronontheweb, do you have any input on this?

For the second question, I really have no answer. I would need to analyze the call chain and see if there are any optimization such as batching upstream of this method, I have a feeling that you're correct, that this method is being called for each produced message, but I would have to confirm it first.

Reply:
So when a `PersistentActor` writes multiple events via `PersistAll` - all of those events end up inside the same `AtomicWrite`; they're considered to be an atomic transaction from the point of view of the actor who wrote them.

The journal is supposed to process all of the `IPersistentRepresentation` (I think) objects that are included in the `AtomicWrite` in a single transaction to preserve atomicity. If there are multiple `AtomicWrite`s to process at once the order doesn't matter since they all came from different actors or the same actor using `PersistAsync` - if writes complete in different orders coming from the same actor I believe the `Resequencer` actor, built into the journal, will unroll the order in which the `WriteMessageSuccess` messages are delivered back to the original actor based on the sequence numbers that were generated prior to the write.

So the amount of things the implementor has to worry about is relatively small - the biggest one is making sure they don't group multiple `AtomicWrite`s into the same transaction. There are some implementations which explicitly break these atomicity guarantees, such as the `BatchingSqlJournal`, but that should be done only while explicitly making a trade off in the name of performance or throughput - not simply because the implementor didn't know better.

Reply:
> For the second question, I really have no answer. I would need to analyze the call chain and see if there are any optimization such as batching upstream of this method, I have a feeling that you're correct, that this method is being called for each produced message, but I would have to confirm it first.

Don't worry about this specific implementation. Worry about the general advice you're to give all implementors about connection management.

Reply:
> That is an interesting question, in theory, there is no guarantee that each `AtomicWrite` Task would be written in order, especially if one of them in the middle failed to be written to the datasource, the ordering would be screwed. @Aaronontheweb, do you have any input on this?
> 
> For the second question, I really have no answer. I would need to analyze the call chain and see if there are any optimization such as batching upstream of this method, I have a feeling that you're correct, that this method is being called for each produced message, but I would have to confirm it first.

From the Docs on asyncwritejournal

Ordering:
>  Journals should aim to persist events in-order for a given `persistenceId`  as otherwise in case of a failure, the persistent state may be end up being inconsistent.

It's not a hard requirement, but it saves a lot of pain down the road, particularly if `persistAsync` is involved.

Atomicity:

> All `PersistentRepr` of the `AtomicWrite` must be written to the data store atomically, i.e. all or none must be stored. If the journal (data store) cannot support atomic writes of multiple events it should reject such writes with a `Try` `Failure` 
> This limitation should also be documented by the journal plugin.

Blocking Sequence Reads on writes:

> Please also note that requests for the highest sequence number may be made concurrently to this call executing for the same `persistenceId`, in particular it is possible that a restarting actor tries to recover before its outstanding writes have completed. In the latter case it is highly desirable to defer reading the highest sequence number until all outstanding writes have completed, otherwise the <see cref="PersistentActor"/> may reuse sequence numbers.

Note there's some edge cases around the above, even if you get it right in the provider itself (e.x. too short a recovery timeout on cluster shards/singletons, and they somehow read before the previous system finished it's somehow-delayed write)

I do know that as far as ordering, [in persistince-jdbc's batching they used `mapAsync` rather than `mapAsyncUnordered`](https://github.com/akka/akka-persistence-jdbc/blob/c89d588108f20646976be10a5028a6dc22958a6e/core/src/main/scala/akka/persistence/jdbc/journal/dao/BaseDao.scala#L22-L33), even though the atomicWrites are atomically grouped (via batchWeighted). I'm not certain whether there's still some edge cases with aggressive PersistAsync usage that could cause issues with this pattern (I think EventSourced handles this with it's batching).

Also, they handle the case of `readHighestSequenceNr` blocking for writes via a Future [Well, here's the ported version since it may be more useful](https://github.com/akkadotnet/Akka.Persistence.Linq2Db/blob/dev/src/Akka.Persistence.Sql.Linq2Db/Journal/Linq2DbWriteJournal.cs#L167-L177).

> If there are multiple AtomicWrites to process at once the order doesn't matter since they all came from different actors or the same actor using PersistAsync

@Aaronontheweb So if you look at the ported code above, what's interesting is that it appears there is an implicit guarantee that for a given call to `WriteMessagesAsync`, all of the `AtomicWrite`s will be the same persistenceId. [here's the scala version for ref](https://github.com/akka/akka-persistence-jdbc/blob/c89d588108f20646976be10a5028a6dc22958a6e/core/src/main/scala/akka/persistence/jdbc/journal/JdbcAsyncWriteJournal.scala#L91-L94).

The implication of all this, is for a given call to `WriteMessagesAsync`, you're best off doing one of the following to keep things as 'Akka' as possible (i.e. fail fast in a recoverable way):

 - Treating all AtomicWrites in that call as a single transacted write to the store
   - very easy to implement for the example IMO
   - As noted above, basically what persistence-jdbc and Persistence.Linq2Db do.
 - Ensuring that all AtomicWrites are chunked into an `AtomicWriteGroup` such that:
   - A chunk contains 1 or more `AtomicWrite`s from the group
   - An `AtomicWrite` is only inside one chunk
   - Chunks are processed in-order
   - As soon as one chunk fails, all subsequent chunks in that Group ignored.

Obviously the second option is a good bit more complicated, but I threw it out there for others who may want to roll their own.